### PR TITLE
Feat(NextjsSite) : extend construct using factory pattern

### LINKED ISSUE #3019: [NextjsSite] Using custom L2 constructs in NextjsSite
First of all thanks a lot for all the impressive work on SST! 

In my current project I have a library of L2 constructs enforcing some configuration for Lambda functions, buckets, distributions, etc. 
I'd like to use them as building blocks inside the NextjsSite.

Have you thought about how the construct can build on top of such custom constructs? 

From what I read, a first possibility would be to extend the NextjsSite and overwrite methods creating resources. This would require to change private methods to protected. It's not great in the long run as the extension would copy some of the logic that you guys have created.

To go a step further, I thought we could use create functions such as https://github.com/serverless-stack/sst/blob/d615080013d8fc09d1bfc93bfe53284866d71c5f/packages/sst/src/constructs/SsrSite.ts#L711-L727 for basic resources (distribution, lambdas, queue). A user that want to use its custom L2 constructs would simply override these protected create methods.

What do you think ?

### PR DESCRIPTION
**Pull request by guillaumeduboc**
resolves #3019 

The `NextjsSite` construct is an L3 construct built on top of cdk's L2 constructs (Function, Bucket, Distribution, Queue).
For teams that have customized their own L2 constructs, it would be great to be able to build a NextjsSite with these base constructs. 

This pr is a proposition to use protected factory method that give possibility to teams to extend `NextjsSite` and use their own customized constructs.
The suggested interface can be found in the tests : https://github.com/serverless-stack/sst/blob/5b8373cfeedab2d5c0fffb5a2081c26cfa491f5f/packages/sst/test/constructs/NextjsSite.test.ts#L156-L169

### CODE REVIEW

File: www/docs/advanced/extending-sst.md [hunk 1]
```diff
@@ -62,6 +62,56 @@ Most of these CDK constructs are exposed as properties of the `Api` construct.
...
+          architecture: Architecture.ARM_64,
+        });
+      }
+    }
+    ```
+
+2. Extend the SST construct
+
+    ```js
+    class CustomNextjsSite extends NextjsSite {
```
[the remark this answers is not in the document]
Comment (khuezy) [reply www/docs/advanced/extending-sst.md:100]: See: https://github.com/serverless-stack/sst/blob/master/packages/sst/src/cli/commands/dev.tsx#L276

[the remark this answers is not in the document]
Comment (guillaumeduboc) [reply www/docs/advanced/extending-sst.md]: good to know ! I edited the example

**Comment by fwang** [comment]
Thanks @guillaumeduboc!

The pattern we've been using is allowing to customize via the `cdk` prop. For example, if there were a way to customize the queue, it would look like: 

```ts
new NextjsSite(stack, "Web", {
  cdk: {
    invalidationQueue, // takes either sqs.QueueProps or sqs.Queue
  }
});
```

Would this pattern work for you?

**Comment by guillaumeduboc** [comment]
Hey @fwang 

I saw that so fare the interface has been the `cdk` prop
I must say that in my use case I'd really like to be able to reuse the constructs we've already define instead of reimplementing everything with props. Furthermore, props are only affecting resources that use the specific props whereas all function created would use the same custom L2 construct in my proposition.

I think it would be nice for sst user that know cdk to have a way to customize in depth their resources.

**Comment by fwang** [comment]
I might be misunderstood something... `cdk` prop will let u reuse the constructs.

For example, in the factory way, to override the bucket:
```ts
class MyNextjsSite extends NextjsSite { 
  protected createS3Bucket() { 
    return new Bucket(this, "S3Bucket", { 
      publicReadAccess: false, 
      blockPublicAccess: BlockPublicAccess.BLOCK_ALL, 
      autoDeleteObjects: true, 
      removalPolicy: RemovalPolicy.DESTROY, 
      ...cdk?.bucket, 
    });
  }
}

new MyNextjsSite(stack, "Web");
```

In the `cdk` way, it would be
```ts
new NextjsSite(stack, "Web", {
  cdk: {
    bucket: { 
      publicReadAccess: false, 
      blockPublicAccess: BlockPublicAccess.BLOCK_ALL, 
      autoDeleteObjects: true, 
      removalPolicy: RemovalPolicy.DESTROY, 
      ...cdk?.bucket, 
    },
  },
});
```

Or this:
```ts
const bucket = new Bucket(this, "S3Bucket", { 
  publicReadAccess: false, 
  blockPublicAccess: BlockPublicAccess.BLOCK_ALL, 
  autoDeleteObjects: true, 
  removalPolicy: RemovalPolicy.DESTROY, 
  ...cdk?.bucket, 
});

new NextjsSite(stack, "Web", {
  cdk: {
    bucket
  },
});
```

The end result from the two approaches are identical right?

**Comment by guillaumeduboc** [comment]
My bad, I didn't understand what you meant

you are right, for the bucket it doesn't change much. And I think this is a great interface for a stateful resource like S3
But for other resources such as lambda functions it's a big difference :

With cdk props : 
First there isn't a prop yet for lambdas. Second It would be convoluted for some of the properties to instantiate the lambda (the `code` for example). Finally we would need 1 prop per function in the construct

With the factory method :
there is no internal logic of sst that propagates outside. The interface grows quickly with minimal impact on the `NextjsSite` codebase

In the end, the difference between the prop and the creator function is that prop would only give the possibility to give 1 instance of 1 resource. Whereas the factory method gives the possibility to create any number of resources, in a controlled way.

**Comment by thdxr** [comment]
We are transition this repo to https://github.com/sst/v2 and are closing this issue for now. If it is still relevant please reopen it there - sorry for the trouble

**Pull request closed without merging**